### PR TITLE
Michelson backend work

### PR DESCRIPTION
**Pull request by cwgoes**
Port old Michelson backend code, now using `morley`.

Follow-up: https://github.com/cryptiumlabs/juvix/issues/168

### CODE REVIEW

**Comment by cwgoes** [comment]
(it does build, CI is just broken right now)

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 1]
```diff
@@ -4,46 +4,53 @@
 module Juvix.Backends.Michelson.Optimisation where
 
 import Juvix.Library
+import Juvix.Backends.Michelson.Compilation.Types
 import Michelson.Typed
 
 {-  This is a simple optimization strategy which replaces sequences of Michelson instructions with equivalent sequences of fewer instructions.
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]: I would probably break these into shorter line prose.

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 2]
```diff
@@ -4,46 +4,53 @@
...
-    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF x y)
-    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_LEFT x y)
-    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_CONS x y)
-    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_NONE x y)
+    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF x y)
+    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_LEFT x y)
+    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_CONS x y)
+    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_NONE x y)
     (DIP e) → DIP |<< optimise' e
+#if defined(OPTIMISE)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]: We could probably replace these CPP flags with some logic

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 3]
```diff
@@ -4,46 +4,53 @@
...
+  (Monad m) ⇒
+  Instr a b →
+  m (Instr a b)
 optimise' expr =
   case expr of
-    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF x y)
-    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_LEFT x y)
-    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_CONS x y)
-    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → return (IF_NONE x y)
+    (IF x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF x y)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]:
```suggestion
    (IF x y) → IF <$> optimise' x <*> optimise' y
```

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 4]
```diff
@@ -4,46 +4,53 @@
...
+    (IF_LEFT x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_LEFT x y)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]:
```suggestion
    (IF_LEFT x y) → IF_LEFT <$> optimise' x <*> optimise' y
```

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 5]
```diff
@@ -4,46 +4,53 @@
...
+    (IF_CONS x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_CONS x y)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]:
```suggestion
    (IF_CONS x y) → IF_CONS <$> optimise' x <*> optimise' y
```

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 6]
```diff
@@ -4,46 +4,53 @@
...
+    (IF_NONE x y) → optimise' x >>= \x → optimise' y >>= \y → pure (IF_NONE x y)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]:
```suggestion
    (IF_NONE x y) → IF_NONE <$> optimise' x <*> optimise' y
```

File: src/Juvix/Backends/Michelson/Optimisation.hs [hunk 7]
```diff
@@ -4,46 +4,53 @@
...
+    (Seq DUP SWAP) → pure DUP
+    (Seq DUP DROP) → pure Nop
+    (Seq SWAP SWAP) → pure Nop
+    (Seq FAILWITH _) → pure FAILWITH
 #endif
     (Seq e Nop) → optimise' e
     (Seq Nop e) → optimise' e
-    (Seq x y) → optimise' x >>= \x → optimise' y >>= \y → return (Seq x y)
-    expr → return expr
+    (Seq x y) → optimise' x >>= \x → optimise' y >>= \y → pure (Seq x y)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Optimisation.hs]:
```suggestion
    (Seq x y) → Seq <$> optimise' x <*> optimise' y
```

File: src/Juvix/Backends/Michelson/Parameterisation.hs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  -- TODO: Add all Michelson instructions which are functions.
+
+  deriving (Show, Eq, Generic)
+
+-- TODO: Add rest of primitive values.
+typeOf ∷ PrimVal → [PrimTy]
+typeOf (PrimConst v) = [PrimTy (M.Type (constType v) "")]
+
+constType ∷ M.Value' Op → M.T
+constType v =
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Parameterisation.hs]: Add an error for the missing cases?

File: src/Juvix/Backends/Michelson/Compilation/Util.hs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  m ExpandedOp
+unpack (Type ty _) binds =
+  case ty of
+    Tbool → do
+      modify @"stack" (drop 1)
+      return (SeqEx [])
+    TPair _ _ fT sT →
+      case binds of
+        [Just fst, Just snd] → do
+          modify @"stack" ((<>) [(VarE fst, fT), (VarE snd, sT)] . drop 1)
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Compilation/Util.hs]: We should make a function to abstract the pattern of `append thing . drop 1`

File: src/Juvix/Backends/Michelson/Compilation/Util.hs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  ( HasState "stack" Stack m,
+    HasThrow "compilationError" CompilationError m
+  ) ⇒
+  [Maybe Symbol] →
+  m ExpandedOp
+unpackDrop binds = genReturn (foldDrop (fromIntegral (length (filter isJust binds))))
+
+position ∷ Symbol → Stack → Maybe Natural
+position _ [] = Nothing
+position n (x : xs) = if fst x == VarE n then Just 0 else (+) 1 |<< position n xs
```
Comment (mariari) [inline src/Juvix/Backends/Michelson/Compilation/Util.hs]:
```suggestion
position n (x : xs) = if fst x == VarE n then Just 0 else succ |<< position n xs
```

File: test/Backends/Michelson.hs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+shouldOptimise instr opt =
+  T.testCase
+    (show instr <> " should optimise to " <> show opt)
+    (optimise' instr T.@=? Just opt)
+
+test_optimise_dup_drop :: T.TestTree
+test_optimise_dup_drop = shouldOptimise (MT.Seq MT.DUP MT.DROP) MT.Nop
+
+test_identity ∷ T.TestTree
+test_identity = shouldCompile identityTerm identityType "parameter unit;storage unit;code {{DUP; {DIP {{}}; {CAR; {NIL operation; {PAIR % %; {DIP {{DROP}}; {}}}}}}}};"
```
Comment (mariari) [inline test/Backends/Michelson.hs]: Nitpick about long line

File: test/Backends/Michelson.hs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+identityContract = M.SomeContract (MT.Seq MT.DUP MT.DROP)
+  (MT.STPair (MT.STList MT.STOperation) MT.STUnit M.-:& M.SNil)
+  (MT.STPair (MT.STList MT.STOperation) MT.STUnit M.-:& M.SNil)
+-}
+
+identityTerm ∷ Term
+identityTerm = J.Lam "x" (J.App (J.App (J.Prim PrimPair) (J.Prim (PrimConst M.ValueNil))) (J.App (J.Prim PrimFst) (J.Var "x")))
+
+identityType ∷ Type
+identityType = J.Pi (J.PrimTy (PrimTy (M.Type (M.TPair "" "" unit unit) ""))) (J.PrimTy (PrimTy (M.Type (M.TPair "" "" opl unit) "")))
```
Comment (mariari) [inline test/Backends/Michelson.hs]: Bit long as well

File: src/Juvix/Backends/Michelson/Compilation/Term.hs [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        pure (M.SeqEx [arg2, arg1, func])
+    -- :: (\a -> b) a ~ (a, s) => (b, s)
+    -- Call-by-value (evaluate argument first).
+    J.App func arg →
+      stackCheck addsOne $ do
+        arg ← termToInstr arg
+        func ← termToInstr func
+        pure (M.SeqEx [arg, func])
+    -- TODO: Linear logic types (when added to core), etc.
+    _ → notYetImplemented
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Compilation/Term.hs]: Yes, it is, I removed it.

File: src/Juvix/Backends/Michelson/Optimisation.hs (hunk 2 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Optimisation.hs]: The goal is to reduce compilation time, so I don't think so... we could switch to the untyped form.

File: src/Juvix/Backends/Michelson/Optimisation.hs (hunk 1 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Optimisation.hs]: Done.

File: src/Juvix/Backends/Michelson/Parameterisation.hs (hunk 8 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Parameterisation.hs]: They will all be filled out later, this function can't fail, I just haven't done it yet.

File: test/Backends/Michelson.hs (hunk 11 above)
Comment (cwgoes) [reply test/Backends/Michelson.hs]: Shortened.

File: test/Backends/Michelson.hs (hunk 12 above)
Comment (cwgoes) [reply test/Backends/Michelson.hs]: Shortened.

File: src/Juvix/Backends/Michelson/Compilation/Term.hs [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  ∀ m.
+  ( HasState "stack" Stack m,
+    HasThrow "compilationError" CompilationError m,
+    HasWriter "compilationLog" [CompilationLog] m
+  ) ⇒
+  Term →
+  M.Type →
+  m Op
+termToMichelson term argTy = do
+  modify @"stack" ((:) (FuncResultE, argTy))
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Compilation/Term.hs]: Updated.

File: src/Juvix/Backends/Michelson/Compilation/Term.hs [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+          -- :: \x -> y ~ (x, s) => (y, s)
+          PrimSnd → stackCheck (lambda 1) $ do
+            genReturn (M.PrimEx (M.CDR "" ""))
+          -- :: \x y -> a ~ (x, (y, s)) => (a, s)
+          PrimPair → stackCheck (lambda 2) $ do
+            modify @"stack" (\((_, xT) : (_, yT) : xs) → (FuncResultE, M.Type (M.TPair "" "" xT yT) "") : xs)
+            pure (M.PrimEx (M.PAIR "" "" "" ""))
+          -- :: a ~ s => (a, s)
+          PrimConst const → stackCheck addsOne $ do
+            case const of
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Compilation/Term.hs]: `stackCheck` doesn't limit matches, it's just a sanity check on the compiler state.

File: src/Juvix/Backends/Michelson/Optimisation.hs (hunk 2 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Optimisation.hs]: I might have to do that anyways to make some optimisations which are semantically correct (mostly with `LAMBDA / EXEC`) but don't typecheck work...

File: src/Juvix/Backends/Michelson/Compilation/Util.hs (hunk 9 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Compilation/Util.hs]: OK, done.

File: src/Juvix/Backends/Michelson/Optimisation.hs (hunk 2 above)
Comment (cwgoes) [reply src/Juvix/Backends/Michelson/Optimisation.hs]: I will do this in a follow-up PR, I expect - ref https://github.com/cryptiumlabs/juvix/issues/168

**Pull request merged**